File: cpp/stack.cpp
#include "singly_linked_list.cpp"

template <typename T>
class Stack{
  public:
    Stack() = default;
    Stack(T firstElem){
      push(firstElem);
    }
    int size(){
      return list.getSize();
    }
    bool isEmpty(){
      return size() == 0;
    }
    void push(T elem){
      list.addLast(elem);
    }

    T pop(){
      if (isEmpty())
        throw std::out_of_range("");
      return list.removeLast();
    }

    T peek(){
      if (isEmpty())
        throw std::out_of_range("");
      return list.peekLast();
    }
  private:
    SinglyLinkedList<T> list;
};

int main(){
  Stack<int> stack;
  for (int i {}; i < 5; i++){
      stack.push(i);
      std::cout<<stack.peek()<<std::endl;
  }
  for (int i {}; i < 3; i++){
      std::cout<<stack.pop()<<std::endl;
  }
  return 0;
}

File: py/stack.py
from __future__ import annotations
from typing import TypeVar
from singly_linked_list import SinglyLinkedList


T = TypeVar('T')


class Stack:
    def __init__(self: Stack, firstElem: T = None):
        if firstElem:
            self.list = SinglyLinkedList()
            self.push(firstElem)
        else:
            self.mlist = SinglyLinkedList()

    def size(self: Stack) -> int:
        return self.list.size

    def isEmpty(self: Stack) -> bool:
        return self.size() == 0

    def push(self: Stack, elem: T) -> None:
        self.list.addLast(elem)

    def pop(self: Stack) -> T:
        if self.isEmpty():
            raise IndexError()
        return self.list.removeLast()

    def peek(self: Stack) -> T:
        if self.isEmpty():
            raise IndexError()
        return self.list.peekLast()


if __name__ == '__main__':
    stack = Stack(5)
    for i in range(5):
        stack.push(i)
        print(stack.peek())
    for i in range(3):
        print(stack.pop())
